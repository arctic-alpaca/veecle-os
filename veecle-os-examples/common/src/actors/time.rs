//! A minimal example using the time abstraction.

use core::convert::Infallible;
use core::fmt::Debug;

use veecle_os::osal::api::time::{Duration, Instant, Interval, TimeAbstraction};
use veecle_os::runtime::{InitializedReader, Storable, Writer};

const INTERVAL_PERIOD: Duration = Duration::from_secs(1);
const ABS_ERROR: Duration = Duration::from_millis(10);

/// A clock tick, which happens in an specific moment in time.
#[derive(Debug, PartialEq, Clone, Storable)]
pub struct Tick {
    at: Instant,
}

/// Emits [`Tick`] every second.
#[veecle_os::runtime::actor]
pub async fn ticker_actor<T>(
    mut tick_writer: Writer<'_, Tick>,
) -> Result<Infallible, veecle_os::osal::api::Error>
where
    T: TimeAbstraction,
{
    let mut interval = T::interval(INTERVAL_PERIOD);

    loop {
        interval.tick().await?;
        tick_writer.write(Tick { at: T::now() }).await;
    }
}

/// Prints every received tick via `telemetry` (if enabled).
/// Additionally, prints to stderr if the time between
/// the previous and current tick differs by more than 10 millis.
#[veecle_os::runtime::actor]
pub async fn ticker_reader(mut reader: InitializedReader<'_, Tick>) -> Infallible {
    let mut previous: Option<Instant> = None;

    loop {
        reader
            .wait_for_update()
            .await
            .read(|&Tick { at: tick_at }| {
                #[cfg(feature = "telemetry")]
                veecle_os::telemetry::info!(
                    "last tick was at",
                    tick_at = alloc::format!("{tick_at:?}")
                );
                if let Some(previous) = previous
                    && let Some(elapsed) = tick_at.duration_since(previous)
                {
                    let _ = elapsed;
                    #[cfg(feature = "telemetry")]
                    veecle_os::telemetry::info!(
                        "since last tick",
                        elapsed = alloc::format!("{elapsed:?}")
                    );
                }
                if previous
                    .replace(tick_at)
                    .and_then(|previous| tick_at.duration_since(previous))
                    .map(|diff| diff.abs_diff(INTERVAL_PERIOD) > ABS_ERROR)
                    .unwrap_or_default()
                {
                    #[cfg(feature = "telemetry")]
                    veecle_os::telemetry::error!(
                        "previous and latest tick differ more than",
                        period = i64::try_from(INTERVAL_PERIOD.as_millis()).unwrap(),
                        error_bound = i64::try_from(ABS_ERROR.as_millis()).unwrap()
                    );
                }
            });
    }
}
