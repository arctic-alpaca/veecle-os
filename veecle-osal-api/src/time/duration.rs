//! This module implements a [`Duration`] with microsecond precision.

use core::fmt;
use core::num::TryFromIntError;
use core::ops::{Add, Div, Mul, Sub};

/// Duration represents a span of time.
///
/// Negative durations are not supported. [`Duration`] is not meant to be used
/// for math operations.
#[derive(
    Default, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, serde::Serialize, serde::Deserialize,
)]
pub struct Duration {
    micros: u64,
}

impl Duration {
    /// The largest value that can be represented by the `Duration` type.
    ///
    /// # Examples
    ///
    /// ```
    /// use veecle_osal_api::time::Duration;
    ///
    /// assert_eq!(Duration::MAX, Duration::from_micros(u64::MAX));
    /// ```
    pub const MAX: Duration = Duration { micros: u64::MAX };

    /// A duration of zero time.
    ///
    /// # Examples
    ///
    /// ```
    /// use veecle_osal_api::time::Duration;
    ///
    /// assert_eq!(Duration::ZERO, Duration::from_micros(0));
    /// ```
    pub const ZERO: Duration = Duration { micros: 0 };

    /// Factor of microseconds per second.
    const MICROS_PER_SECOND: u64 = 1_000_000;
    /// Factor of milliseconds per second.
    const MILLIS_PER_SECOND: u64 = 1_000;

    /// Creates a duration from the specified number of seconds.
    ///
    /// # Examples
    ///
    /// ```
    /// use veecle_osal_api::time::Duration;
    ///
    /// assert_eq!(Duration::from_secs(1), Duration::from_millis(1000));
    /// ```
    pub const fn from_secs(secs: u64) -> Duration {
        Duration {
            micros: secs * Self::MICROS_PER_SECOND,
        }
    }

    /// Creates a duration from the specified number of milliseconds.
    ///
    /// # Examples
    ///
    /// ```
    /// use veecle_osal_api::time::Duration;
    ///
    /// assert_eq!(Duration::from_secs(1), Duration::from_millis(1000));
    /// ```
    pub const fn from_millis(millis: u64) -> Duration {
        Duration {
            micros: millis * Self::MILLIS_PER_SECOND,
        }
    }

    /// Creates a duration from the specified number of microseconds.
    ///
    /// # Examples
    ///
    /// ```
    /// use veecle_osal_api::time::Duration;
    ///
    /// assert_eq!(Duration::from_secs(1), Duration::from_micros(1000000));
    /// ```
    pub const fn from_micros(micros: u64) -> Duration {
        Duration { micros }
    }

    /// Returns the total amount of seconds, rounded down.
    ///
    /// # Examples
    ///
    /// ```
    /// use veecle_osal_api::time::Duration;
    ///
    /// assert_eq!(Duration::from_millis(1980).as_secs(), 1);
    /// ```
    pub const fn as_secs(&self) -> u64 {
        self.micros / Self::MICROS_PER_SECOND
    }

    /// Returns the total amount of milliseconds, rounded down.
    ///
    /// # Examples
    ///
    /// ```
    /// use veecle_osal_api::time::Duration;
    ///
    /// assert_eq!(Duration::from_millis(1980).as_millis(), 1980);
    /// ```
    pub const fn as_millis(&self) -> u64 {
        self.micros / Self::MILLIS_PER_SECOND
    }

    /// Returns the total amount of microseconds.
    ///
    /// # Examples
    ///
    /// ```
    /// use veecle_osal_api::time::Duration;
    ///
    /// assert_eq!(Duration::from_millis(1980).as_micros(), 1980000);
    /// ```
    pub const fn as_micros(&self) -> u64 {
        self.micros
    }

    /// Adds one Duration to another, returning a new Duration or None in the event of an overflow.
    ///
    /// # Examples
    ///
    /// ```
    /// use veecle_osal_api::time::Duration;
    ///
    /// assert_eq!(
    ///     Duration::from_secs(1).checked_add(Duration::from_secs(1)),
    ///     Some(Duration::from_secs(2))
    /// );
    /// assert_eq!(Duration::MAX.checked_add(Duration::from_secs(1)), None);
    /// ```
    pub fn checked_add(self, rhs: Duration) -> Option<Duration> {
        self.micros
            .checked_add(rhs.micros)
            .map(|micros| Duration { micros })
    }

    /// Subtracts one Duration from another, returning a new Duration or None in the event of an underflow.
    ///
    /// # Examples
    ///
    /// ```
    /// use veecle_osal_api::time::Duration;
    ///
    /// assert_eq!(
    ///     Duration::from_secs(2).checked_sub(Duration::from_secs(1)),
    ///     Some(Duration::from_secs(1))
    /// );
    /// assert_eq!(Duration::from_secs(1).checked_sub(Duration::from_secs(2)), None);
    /// ```
    pub fn checked_sub(self, rhs: Duration) -> Option<Duration> {
        self.micros
            .checked_sub(rhs.micros)
            .map(|micros| Duration { micros })
    }

    /// Multiplies one Duration by a scalar `u32`, returning a new Duration or None in the event of an overflow.
    ///
    /// # Examples
    ///
    /// ```
    /// use veecle_osal_api::time::Duration;
    ///
    /// assert_eq!(Duration::from_secs(1).checked_mul(2), Some(Duration::from_secs(2)));
    /// assert_eq!(Duration::MAX.checked_mul(2), None);
    /// ```
    pub fn checked_mul(self, rhs: u32) -> Option<Duration> {
        self.micros
            .checked_mul(rhs as _)
            .map(|micros| Duration { micros })
    }

    /// Divides one Duration by a scalar `u32`, returning a new Duration or None if `rhs == 0`.
    ///
    /// # Examples
    ///
    /// ```
    /// use veecle_osal_api::time::Duration;
    ///
    /// assert_eq!(Duration::from_secs(1).checked_div(2), Some(Duration::from_millis(500)));
    /// assert_eq!(Duration::from_secs(1).checked_div(0), None);
    /// ```
    pub fn checked_div(self, rhs: u32) -> Option<Duration> {
        self.micros
            .checked_div(rhs as _)
            .map(|micros| Duration { micros })
    }

    /// Returns the absolute difference between self and rhs.
    ///
    /// # Examples
    ///
    /// ```
    /// use veecle_osal_api::time::Duration;
    ///
    /// assert_eq!(Duration::from_secs(1).abs_diff(Duration::from_secs(2)), Duration::from_secs(1));
    /// assert_eq!(Duration::from_secs(2).abs_diff(Duration::from_secs(1)), Duration::from_secs(1));
    /// ```
    pub fn abs_diff(self, rhs: Self) -> Duration {
        Self {
            micros: self.micros.abs_diff(rhs.micros),
        }
    }

    /// Returns a duration of approximately 50 years.
    ///
    /// No leap-seconds/days have been taken into account.
    ///
    /// Can be used in place of [MAX][Self::MAX] to avoid overflows.
    pub(crate) fn max_no_overflow_alias() -> Self {
        // seconds * minutes * hours * days * weeks * years
        Self::from_secs(60 * 60 * 24 * 7 * 52 * 50)
    }
}

impl Add for Duration {
    type Output = Self;

    /// # Panics
    ///
    /// This function may panic if the resulting duration overflows. See [`Duration::checked_add`] for a version
    /// without panic.
    ///
    /// # Examples
    ///
    /// ```
    /// use veecle_osal_api::time::Duration;
    ///
    /// assert_eq!(Duration::from_secs(1) + Duration::from_secs(1), Duration::from_secs(2));
    /// ```
    ///
    /// ```should_panic
    /// use veecle_osal_api::time::Duration;
    ///
    /// let _ = Duration::MAX + Duration::from_secs(1);
    /// ```
    fn add(self, rhs: Self) -> Self::Output {
        let Some(result) = self.checked_add(rhs) else {
            panic!("overflow when adding two durations");
        };

        result
    }
}

impl Sub for Duration {
    type Output = Self;

    /// # Panics
    ///
    /// This function may panic if the resulting duration underflows. See [`Duration::checked_sub`] for a
    /// version without panic.
    ///
    /// # Examples
    ///
    /// ```
    /// use veecle_osal_api::time::Duration;
    ///
    /// assert_eq!(Duration::from_secs(2) - Duration::from_secs(1), Duration::from_secs(1));
    /// ```
    ///
    /// ```should_panic
    /// use veecle_osal_api::time::Duration;
    ///
    /// let _ = Duration::from_secs(1) - Duration::from_secs(2);
    /// ```
    fn sub(self, rhs: Self) -> Self::Output {
        let Some(result) = self.checked_sub(rhs) else {
            panic!("underflow when subtracting two durations");
        };

        result
    }
}

impl Mul<u32> for Duration {
    type Output = Self;

    /// # Panics
    ///
    /// This function may panic if the resulting duration overflows. See [`Duration::checked_mul`] for a version
    /// without panic.
    ///
    /// # Examples
    ///
    /// ```
    /// use veecle_osal_api::time::Duration;
    ///
    /// assert_eq!(Duration::from_secs(1) * 2, Duration::from_secs(2));
    /// ```
    ///
    /// ```should_panic
    /// use veecle_osal_api::time::Duration;
    ///
    /// let _ = Duration::MAX * 2;
    /// ```
    fn mul(self, rhs: u32) -> Self::Output {
        let Some(result) = self.checked_mul(rhs) else {
            panic!("overflow when multiplying a duration by a scalar");
        };

        result
    }
}

impl Mul<Duration> for u32 {
    type Output = Duration;

    /// # Panics
    ///
    /// This function may panic if the resulting duration overflows. See [`Duration::checked_mul`] for a version
    /// without panic.
    ///
    /// # Examples
    ///
    /// ```
    /// use veecle_osal_api::time::Duration;
    ///
    /// assert_eq!(2 * Duration::from_secs(1), Duration::from_secs(2));
    /// ```
    ///
    /// ```should_panic
    /// use veecle_osal_api::time::Duration;
    ///
    /// let _ = 2 * Duration::MAX;
    /// ```
    fn mul(self, rhs: Duration) -> Self::Output {
        rhs * self
    }
}

impl Div<u32> for Duration {
    type Output = Self;

    /// # Panics
    ///
    /// This function may panic if the duration is divided by zero. See [`Duration::checked_div`] for a version
    /// without panic.
    ///
    /// # Examples
    ///
    /// ```
    /// use veecle_osal_api::time::Duration;
    ///
    /// assert_eq!(Duration::from_secs(1) / 2, Duration::from_millis(500));
    /// ```
    ///
    /// ```should_panic
    /// use veecle_osal_api::time::Duration;
    ///
    /// let _ = Duration::from_secs(1) / 0;
    /// ```
    fn div(self, rhs: u32) -> Self::Output {
        let Some(result) = self.checked_div(rhs) else {
            panic!("divided a duration by zero");
        };

        result
    }
}

impl fmt::Debug for Duration {
    /// # Examples
    ///
    /// ```
    /// use veecle_osal_api::time::Duration;
    ///
    /// let duration = Duration::from_millis(1980);
    /// assert_eq!(format!("{duration:?}"), "1s.980000us");
    /// ```
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(
            f,
            "{}s.{}us",
            self.as_secs(),
            self.as_micros() % Self::MICROS_PER_SECOND
        )
    }
}

impl TryFrom<core::time::Duration> for Duration {
    type Error = TryFromIntError;

    /// # Examples
    ///
    /// ```
    /// use veecle_osal_api::time::Duration;
    ///
    /// assert_eq!(Duration::try_from(core::time::Duration::from_secs(1)), Ok(Duration::from_secs(1)));
    /// ```
    fn try_from(value: core::time::Duration) -> Result<Self, Self::Error> {
        value.as_micros().try_into().map(Self::from_micros)
    }
}

impl From<Duration> for core::time::Duration {
    /// # Examples
    ///
    /// ```
    /// use veecle_osal_api::time::Duration;
    ///
    /// assert_eq!(
    ///     core::time::Duration::from(Duration::from_secs(1)),
    ///     core::time::Duration::from_secs(1)
    /// );
    /// ```
    fn from(value: Duration) -> Self {
        Self::from_micros(value.as_micros())
    }
}
