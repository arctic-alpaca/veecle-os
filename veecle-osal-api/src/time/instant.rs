//! This module implements an [`Instant`] with microsecond precision.

use core::fmt;
use core::num::NonZeroU64;
use core::ops::{Add, Sub};

use super::Duration;

/// An Instant in time. Instants should be always increasing and are
/// generally obtainable through the operating system time driver.
#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub struct Instant {
    micros: NonZeroU64,
}

impl Instant {
    /// The largest value that can be represented by the [`Instant`] type.
    ///
    /// # Examples
    ///
    /// ```
    /// use veecle_osal_api::time::{Duration, Instant, TimeAbstraction};
    /// use veecle_osal_std::time::Time;
    ///
    /// let now = Time::now();
    ///
    /// assert!(Instant::MAX > now);
    /// ```
    pub const MAX: Instant = Instant {
        micros: NonZeroU64::MAX,
    };

    /// The smallest value that can be represented by the [`Instant`] type.
    ///
    /// # Examples
    ///
    /// ```
    /// use veecle_osal_api::time::{Duration, Instant, TimeAbstraction};
    /// use veecle_osal_std::time::Time;
    ///
    /// let now = Time::now();
    ///
    /// assert!(Instant::MIN <= now);
    /// ```
    pub const MIN: Instant = Instant {
        micros: NonZeroU64::MIN,
    };

    /// Returns the [`Duration`] between this [`Instant`] and the give one if, and only if,
    /// the given one is earlier than this, otherwise returns None.
    ///
    /// # Examples
    ///
    /// ```
    /// use veecle_osal_api::time::{Duration, Instant, TimeAbstraction};
    /// use veecle_osal_std::time::Time;
    ///
    /// let begin = Time::now();
    ///
    /// std::thread::sleep(core::time::Duration::from_millis(1));
    ///
    /// let end = Time::now();
    /// assert!(end.duration_since(begin).unwrap() > Duration::ZERO);
    /// assert_eq!(begin.duration_since(end), None);
    /// ```
    pub const fn duration_since(&self, earlier: Instant) -> Option<Duration> {
        if self.micros.get() < earlier.micros.get() {
            return None;
        }

        Some(Duration::from_micros(
            self.micros.get() - earlier.micros.get(),
        ))
    }

    /// Adds one [`Duration`] to self, returning a new [`Instant`] or None in the event of an overflow.
    ///
    /// # Examples
    ///
    /// ```
    /// use veecle_osal_api::time::{Duration, Instant, TimeAbstraction};
    /// use veecle_osal_std::time::Time;
    ///
    /// let now = Time::now();
    ///
    /// assert!(now.checked_add(Duration::from_secs(1)).unwrap() > now);
    /// assert_eq!(now.checked_add(Duration::MAX), None);
    /// ```
    pub fn checked_add(self, rhs: Duration) -> Option<Instant> {
        self.micros
            .checked_add(rhs.as_micros())
            .map(|micros| Instant { micros })
    }

    /// Subs one [`Duration`] from self, returning a new [`Instant`] or None in the event of an underflow.
    ///
    /// # Examples
    ///
    /// ```
    /// use veecle_osal_api::time::{Duration, Instant, TimeAbstraction};
    /// use veecle_osal_std::time::Time;
    ///
    /// let begin = Time::now();
    ///
    /// std::thread::sleep(core::time::Duration::from_millis(1));
    ///
    /// let end = Time::now();
    ///
    /// assert!(end.checked_sub(Duration::from_millis(1)).unwrap() < end);
    /// assert_eq!(end.checked_sub(Duration::MAX), None);
    /// ```
    pub fn checked_sub(&self, rhs: Duration) -> Option<Instant> {
        self.micros
            .get()
            .checked_sub(rhs.as_micros())
            .and_then(NonZeroU64::new)
            .map(|micros| Instant { micros })
    }
}

impl Add<Duration> for Instant {
    type Output = Self;

    /// # Panics
    ///
    /// This function may panic if the resulting instant overflows. See [`Instant::checked_add`] for a version
    /// without panic.
    ///
    /// # Examples
    ///
    /// ```
    /// use veecle_osal_api::time::{Duration, Instant, TimeAbstraction};
    /// use veecle_osal_std::time::Time;
    ///
    /// let now = Time::now();
    ///
    /// assert!(now + Duration::from_secs(1) > now);
    /// ```
    ///
    /// ```should_panic
    /// use veecle_osal_api::time::{Duration, Instant, TimeAbstraction};
    /// use veecle_osal_std::time::Time;
    ///
    /// let now = Time::now();
    ///
    /// let _ = now + Duration::MAX;
    /// ```
    fn add(self, rhs: Duration) -> Self::Output {
        let Some(result) = self.checked_add(rhs) else {
            panic!("overflow when adding a duration to an instant");
        };

        result
    }
}

impl Sub<Duration> for Instant {
    type Output = Self;

    /// # Panics
    ///
    /// This function may panic if the resulting instant underflows. See [`Instant::checked_sub`] for a
    /// version without panic.
    ///
    /// # Examples
    ///
    /// ```
    /// use veecle_osal_api::time::{Duration, Instant, TimeAbstraction};
    /// use veecle_osal_std::time::Time;
    ///
    /// let begin = Time::now();
    ///
    /// std::thread::sleep(core::time::Duration::from_millis(1));
    ///
    /// let end = Time::now();
    ///
    /// assert!(end - Duration::from_millis(1) < end);
    /// ```
    ///
    /// ```should_panic
    /// use veecle_osal_api::time::{Duration, Instant, TimeAbstraction};
    /// use veecle_osal_std::time::Time;
    ///
    /// let now = Time::now();
    ///
    /// let _ = now - Duration::MAX;
    /// ```
    fn sub(self, rhs: Duration) -> Self::Output {
        let Some(result) = self.checked_sub(rhs) else {
            panic!("underflow when subtracting a duration from an instant");
        };

        result
    }
}

impl fmt::Debug for Instant {
    /// # Examples
    ///
    /// ```
    /// use veecle_osal_api::time::{Duration, Instant, TimeAbstraction};
    ///
    /// let instant = Instant::MIN + Duration::from_millis(1980);
    /// assert_eq!(format!("{instant:?}"), "1s.980000us");
    /// ```
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        self.duration_since(Self::MIN)
            .expect("instant should be at least Instant::MIN")
            .fmt(f)
    }
}
