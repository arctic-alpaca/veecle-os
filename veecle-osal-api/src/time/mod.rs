//! Abstractions for time-based operations.
//!
//! The main purpose of this module is to provide the definition of [`TimeAbstraction`], the trait that has to be
//! implemented to interact with the underlying operating system when performing time-based operations.
//! In order to keep the time abstractions as decoupled as possible from the running environment, this module provides
//! its own [`Instant`] and [`Duration`] types.
//!
//! # Example
//!
//! Runtime actors using time-based operations should rely on the [`TimeAbstraction`] trait, and never use specific
//! implementations. It is during the runtime's setup that the concrete implementation for the targeted environment has
//! to be specified.
//!
//! ```rust
//! use core::convert::Infallible;
//!
//! use veecle_osal_api::time::{Duration, TimeAbstraction};
//! use veecle_osal_std::time::Time;
//! use veecle_os_runtime::{Reader, Storable, Writer};
//!
//! #[derive(Debug, Clone, PartialEq, Eq, Default, Storable)]
//! pub struct Tick {
//!     since_epoch: Duration,
//! }
//!
//! #[veecle_os_runtime::actor]
//! async fn tick_writer<T>(mut writer: Writer<'_, Tick>) -> Infallible
//! where
//!     T: TimeAbstraction,
//! {
//!     let epoch = T::now();
//!
//!     loop {
//!         let _ = T::sleep_until(T::now() + Duration::from_secs(1)).await;
//!         writer
//!             .write(Tick {
//!                 since_epoch: T::now()
//!                     .duration_since(epoch)
//!                     .expect("now should be later than epoch"),
//!             })
//!             .await;
//!     }
//! }
//!
//! #[veecle_os_runtime::actor]
//! async fn tick_reader(mut reader: Reader<'_, Tick>) -> Infallible {
//!     loop {
//!         reader.wait_for_update().await.read(|tick| {
//!             println!("[READER TASK] Tick received: {tick:?}");
//! #           // Exit the application to allow doc-tests to complete.
//! #           std::process::exit(0);
//!         })
//!     }
//! }
//!
//! # let mut rt = tokio::runtime::Runtime::new().unwrap();
//! # rt.block_on(async move {
//! #
//! veecle_os_runtime::execute! {
//!     store: [Tick],
//!     actors: [
//!         TickWriter<Time>,
//!         TickReader,
//!     ]
//! }.await;
//!
//! unreachable!("the runtime instance does not return");
//! # })
//! ```

#![allow(async_fn_in_trait)] // Auto-bounds are not necessary here.

use core::future::IntoFuture;

use futures::future::Either;

mod duration;
mod instant;
mod system_time;
mod timeout;

pub use self::duration::Duration;
pub use self::instant::Instant;
pub use self::system_time::{SystemTime, SystemTimeError, SystemTimeSync};
pub use self::timeout::Exceeded;
use crate::Error;

/// A stream of periodic ticks, created by [`TimeAbstraction::interval`].
pub trait Interval {
    /// Completes when the next period has been reached (unless there is an error).
    ///
    /// If the stream consumer falls behind and multiple periods go by between reading from the stream, the stream will
    /// keep track of the missed periods and instantly yield them until caught up.
    async fn tick(&mut self) -> Result<(), Error>;
}

/// `TimeAbstraction` is used to perform time-related operations in a platform-agnostic manner.
pub trait TimeAbstraction {
    /// Retrieves the current time.
    fn now() -> Instant;

    /// Returns a future that resolves successfully at the specified `deadline` (or earlier with an error).
    async fn sleep_until(deadline: Instant) -> Result<(), Error>;

    /// Returns a future that resolves successfully after the specified `duration` (or earlier with an error).
    ///
    /// If the `duration` overflows `Instant`, the method sleeps for an unspecified time.
    async fn sleep(duration: Duration) -> Result<(), Error> {
        match Self::now().checked_add(duration) {
            Some(deadline) => Self::sleep_until(deadline).await,
            None => Self::sleep_until(Self::now() + Duration::max_no_overflow_alias()).await,
        }
    }

    /// Returns a future that will resolve when: the wrapped future resolves, the `deadline` is reached, or there is an
    /// error.
    async fn timeout_at<F>(
        deadline: Instant,
        future: F,
    ) -> Result<F::Output, Either<Exceeded, Error>>
    where
        Self: Sized,
        F: IntoFuture,
    {
        self::timeout::timeout_at::<Self, _>(deadline, future.into_future()).await
    }

    /// Returns an [`Interval`] that will yield an item straight away and then once every `period` (unless there is an error).
    ///
    /// If the stream consumer falls behind and multiple periods go by between reading from the stream, the stream will
    /// keep track of the missed periods and instantly yield them until caught up.
    #[must_use]
    fn interval(period: Duration) -> impl Interval;
}
