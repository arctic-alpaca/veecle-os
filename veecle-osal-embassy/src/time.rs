//! Time related system utilities.

use core::sync::atomic::{AtomicU32, Ordering};
use veecle_osal_api::Error;
pub use veecle_osal_api::time::{
    Duration, Instant, Interval, SystemTime, SystemTimeError, SystemTimeSync, TimeAbstraction,
};

/// Implements the [`TimeAbstraction`] trait for embassy.
///
/// ## Details
///
/// - Before using [`Self::duration_since_epoch`], you are expected to synchronize system time via
///   [`Self::set_system_time`].
/// - Maximum precision of system time synchronization is limited to seconds.
#[derive(Debug)]
pub struct Time;

impl TimeAbstraction for Time {
    fn now() -> Instant {
        Instant::MIN + Duration::from_millis(embassy_time::Instant::now().as_millis())
    }

    async fn sleep_until(deadline: Instant) -> Result<(), Error> {
        Self::sleep(
            deadline
                .duration_since(Self::now())
                .unwrap_or(Duration::ZERO),
        )
        .await
    }

    async fn sleep(duration: Duration) -> Result<(), Error> {
        // `embassy_time::Timer::after` panics on overflow, so we reduce the duration until it doesn't overflow.
        // This does only reduce the chance of overflow and thus a panic.
        // We cannot prevent all situations in which an overflow might occur.
        let mut duration = duration;
        while Self::now().checked_add(duration).is_none() {
            duration = duration / 2;
        }
        embassy_time::Timer::after(embassy_time::Duration::from_millis(duration.as_millis())).await;
        Ok(())
    }

    fn interval(period: Duration) -> impl Interval
    where
        Self: Sized,
    {
        struct IntervalInternal {
            ticker: embassy_time::Ticker,
            first_poll: bool,
        }

        impl Interval for IntervalInternal {
            async fn tick(&mut self) -> Result<(), Error> {
                // Embassy's ticker doesn't immediately yield an item, so we need to do that to conform to the trait contract.
                if self.first_poll {
                    self.first_poll = false;
                    return Ok(());
                }

                self.ticker.next().await;
                Ok(())
            }
        }

        IntervalInternal {
            ticker: embassy_time::Ticker::every(embassy_time::Duration::from_millis(
                period.as_millis(),
            )),
            first_poll: true,
        }
    }
}

static SYSTEM_TIME_OFFSET_SECONDS: AtomicU32 = AtomicU32::new(0);

impl SystemTime for Time {
    fn duration_since_epoch() -> Result<Duration, SystemTimeError> {
        let offset = SYSTEM_TIME_OFFSET_SECONDS.load(Ordering::Relaxed);
        if offset == 0 {
            Err(SystemTimeError::Unsynchronized)
        } else {
            let duration_since_start = Self::now()
                .duration_since(Instant::MIN)
                .expect("now can't be less than min time");
            Ok(duration_since_start + Duration::from_secs(offset as u64))
        }
    }
}

impl SystemTimeSync for Time {
    fn set_system_time(duration_since_epoch: Duration) -> Result<(), SystemTimeError> {
        let duration_since_start = Self::now()
            .duration_since(Instant::MIN)
            .expect("now can't be less than min time");

        if duration_since_epoch < duration_since_start {
            Err(SystemTimeError::EpochIsLaterThanStartTime)
        } else {
            SYSTEM_TIME_OFFSET_SECONDS.store(
                (duration_since_epoch - duration_since_start).as_secs() as u32,
                Ordering::Relaxed,
            );
            Ok(())
        }
    }
}
