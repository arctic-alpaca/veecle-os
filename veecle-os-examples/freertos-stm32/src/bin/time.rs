#![no_std]
#![no_main]
#![allow(non_snake_case)]

use core::ffi::c_void;

use cortex_m::asm;
use cortex_m_rt::{ExceptionFrame, entry, exception};
use stm32f7xx_hal as _;
use veecle_freertos_integration::*;
use veecle_os::osal::freertos::log::{Log, LogTarget};
use veecle_os::osal::freertos::time::Time;
use veecle_os_examples_common::actors::time::{Tick, TickerActor, TickerReader};

extern crate panic_halt;

// SAFETY: We don't use any non-FreeRTOS threads.
#[global_allocator]
static GLOBAL: FreeRtosAllocator = unsafe { FreeRtosAllocator::new() };

#[entry]
fn main() -> ! {
    // TODO(DEV-533/DEV-559): Initialize `veecle_os::telemetry` here, and use it for later `println`s.

    Log::println(format_args!("Configuring tasks"));

    veecle_freertos_integration::hooks::set_on_assert(|file_name, line| {
        Log::println(format_args!("{file_name}: {line}"));
    });

    Task::new()
        .name(c"time example")
        .stack_size(4096)
        .priority(TaskPriority(2))
        .start(|_| {
            veecle_os::osal::freertos::task::block_on_future(veecle_os::runtime::execute! {
                store: [Tick],
                actors: [
                    TickerReader,
                    TickerActor<Time>,
                ],
            });
        })
        .unwrap();

    veecle_freertos_integration::scheduler::start_scheduler();
    panic!("FreeRTOS scheduler should not return");
}

#[exception]
unsafe fn DefaultHandler(_irqn: i16) {}

#[exception]
#[allow(clippy::empty_loop)]
unsafe fn HardFault(_ef: &ExceptionFrame) -> ! {
    asm::bkpt();
    loop {}
}

// SAFETY: `vApplicationStackOverflowHook` matches the signature in `task.h` of the FreeRTOS kernel source.
// FreeRTOS itself doesn't provide a definition for this function.
// Neither does this example, which means this is the only instance of this function being exported.
#[unsafe(no_mangle)]
fn vApplicationStackOverflowHook(_pxTask: *const c_void, _pcTaskName: *const u8) {
    asm::bkpt();
}
