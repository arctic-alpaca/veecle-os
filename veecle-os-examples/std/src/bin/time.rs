use veecle_os::osal::std::time::Time;
use veecle_os_examples_common::actors::time::{Tick, TickerActor, TickerReader};

#[veecle_os::osal::std::main(telemetry = true)]
async fn main() {
    veecle_os::runtime::execute! {
        store: [Tick],
        actors: [
            TickerReader,
            TickerActor<Time>,
        ],
    }
    .await;
}
