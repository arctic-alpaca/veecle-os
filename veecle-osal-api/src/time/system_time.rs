//! Provides system time functionality with support for time synchronization.

use crate::time::Duration;

/// Provides a measurement of the system time.
pub trait SystemTime {
    /// Calculates the [`Duration`] elapsed since epoch.
    ///
    /// ## Notes
    ///
    /// OSAL implementations might require time to be synchronized by using the [`SystemTimeSync`] before using this
    /// function. Review the documentation on the implementation you are using to learn platform-specific information.
    ///
    /// ## Errors
    ///
    /// - [`SystemTimeError::Unsynchronized`] when system was not synchronized previously.
    fn duration_since_epoch() -> Result<Duration, SystemTimeError>;
}

/// Allows manual synchronization of [`SystemTime`].
pub trait SystemTimeSync {
    /// Updates the current system time.
    ///
    /// Use this to synchronize with real-world clock (e.g., provide time obtained from NTP).
    ///
    /// ## Errors
    ///
    /// - [`SystemTimeError::EpochIsLaterThanStartTime`] when the provided duration is less than the program's execution
    ///   time.
    fn set_system_time(elapsed_from_epoch: Duration) -> Result<(), SystemTimeError>;
}

/// An error that may happen while working with System Time.
#[derive(Debug, Eq, PartialEq)]
pub enum SystemTimeError {
    /// Occurs when an attempt is made to get system time, but it was not synchronized earlier.
    Unsynchronized,
    /// Occurs when an attempt is made to synchronize using an epoch time that is later than the program's start time.
    EpochIsLaterThanStartTime,
}

impl core::error::Error for SystemTimeError {
    fn source(&self) -> Option<&(dyn core::error::Error + 'static)> {
        match self {
            SystemTimeError::Unsynchronized => None,
            SystemTimeError::EpochIsLaterThanStartTime => None,
        }
    }
}

impl core::fmt::Display for SystemTimeError {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        match self {
            SystemTimeError::Unsynchronized => write!(f, "{self:?}"),
            SystemTimeError::EpochIsLaterThanStartTime => write!(f, "{self:?}"),
        }
    }
}
