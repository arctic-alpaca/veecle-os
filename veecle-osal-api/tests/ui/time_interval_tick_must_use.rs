#![deny(unused_must_use)]

use veecle_osal_api::time::Interval;

pub async fn test(mut interval: impl Interval) {
    interval.tick().await;
}

fn main() {}
