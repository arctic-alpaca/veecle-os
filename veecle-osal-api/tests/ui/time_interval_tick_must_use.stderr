error: unused `Result` that must be used
 --> tests/ui/time_interval_tick_must_use.rs:6:5
  |
6 |     interval.tick().await;
  |     ^^^^^^^^^^^^^^^^^^^^^
  |
  = note: this `Result` may be an `Err` variant, which should be handled
note: the lint level is defined here
 --> tests/ui/time_interval_tick_must_use.rs:1:9
  |
1 | #![deny(unused_must_use)]
  |         ^^^^^^^^^^^^^^^
help: use `let _ = ...` to ignore the resulting value
  |
6 |     let _ = interval.tick().await;
  |     +++++++
